### PR TITLE
Refactor Composite Mark Infrastructure

### PR DESCRIPTION
**Pull request by kanitw**
Fix https://github.com/vega/vega-lite/issues/3398 and address part of https://github.com/vega/vega-lite/issues/3443

@starry97 @chanwutk  -- I'm refactoring this so it becomes more what obvious what to do in your PRs.

cc: @mattwchun

### CODE REVIEW

**Comment by kanitw** [comment]
@chanwutk @starry97 -- Please help review this PR + rebase your branch based on this branch and revise your code based on these changes.  

Reviewing changes commit-by-commit will be way easier and help you fix similar things in your PR.

Hopefully this infrastructure will make our code way more consistent.  

For @starry97, this means
1) rebase `2.x-annotation` onto `2.x-composite-mark`
2) then rebase your callout branch onto rebased `2.x-annotation`

File: src/compositemark/boxplot.ts [hunk 1]
```diff
@@ -38,47 +37,54 @@ export interface BoxPlotDef {
...
+export interface BoxPlotPartsMixins {
+  /**
+   * Mark properties for the rect mark that represents the box in a boxplot.
+   */
+  box?: MarkConfig;
+
+  /**
+   * Mark properties for the tick mark that represents the median value in a boxplot.
+   */
+  mid?: MarkConfig;
```
Comment (kanitw) [inline src/compositemark/boxplot.ts]: Anyone has better name than mid?

File: src/compositemark/boxplot.ts [hunk 2]
```diff
@@ -2,86 +2,66 @@ import {isNumber} from 'vega-util';
...
 
-export interface BoxPlotDef {
-  /**
-   * Type of the mark.  For box plots, this should always be `"box-plot"`.
-   * [boxplot](compositemark.html#boxplot)
-   */
-  type: BOXPLOT;
+const BOXPLOT_PART_INDEX: Flag<BoxPlotPart> = {
+  box: 1,
+  mid: 1,
```
Comment (kanitw) [inline src/compositemark/boxplot.ts]: Anyone has any better name for the mid tick?  

I don't want to make it two words `mid-tick`, but `mid` as a part name is also not very satisfying.

**Comment by kanitw** [comment]
Found one bug that needs to be fixed in a separate PR.  `size` in mark def is not always wired correctly yet.

File: src/compositemark/boxplot.ts [hunk 3]
```diff
@@ -2,86 +2,66 @@ import {isNumber} from 'vega-util';
...
 import {GenericUnitSpec, LayerSpec} from './../spec';
 import {Orient} from './../vega.schema';
-import {getMarkSpecificConfigMixins} from './common';
+import {getMarkDefMixins} from './common';
 
 
-export const BOXPLOT: 'box-plot' = 'box-plot';
-export type BOXPLOT = typeof BOXPLOT;
-export type BoxPlotStyle = 'boxWhisker' | 'box' | 'boxMid';
+export const BOXPLOT: 'boxplot' = 'boxplot';
```
Comment (mattwchun) [inline src/compositemark/boxplot.ts]: Why is this changing to `boxplot`? I believe `box-plot` was a decision decision we came up with Jeff. Also box plot is two words and if we use `boxplot` then it'll be inconsistent with `error-bar`

Comment (kanitw) [reply src/compositemark/boxplot.ts]: I want config to be easily referred.  `config["box-plot"]` is awkward.     

`config.boxplot/errorbar` are better.

File: src/compositemark/boxplot.ts (hunk 2 above)
Comment (kanitw) [reply src/compositemark/boxplot.ts]: @mattwchun `center` is perhaps better?

**Comment by kanitw** [comment]
@starry97  I'll merge this for now to unblock @chanwutk then.  If you have any follow up comments, please feel free to comment (or even better, submit a PR)

**Pull request merged**